torch.sgn for complex tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45069 Disable inplace abs for complex tensors
* **#39955 torch.sgn for complex tensors**

resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

also updates the backward definition of `torch.div`, `torch.abs`
Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

anjali411 committed:
torch.sgn for complex tensors

[ghstack-poisoned]

anjali411 commented:
cc. @dylanbespalko

mruberry commented on aten/src/ATen/native/cpu/zmath_std.h:
```diff
@@ -43,6 +43,18 @@ inline double zabs <std::complex<double>, double> (std::complex<double> z) {
   return std::abs(z);
 }
 
+// template<>
```
Commented out? Did you meant to delete this?

anjali411 replied:
yeah updated

mruberry commented on aten/src/ATen/native/cuda/UnarySignKernels.cu:
```diff
@@ -45,8 +45,22 @@ void sign_kernel_cuda(TensorIterator& iter){
+  return v.sgn();
+}
+
+void sign_kernel_cuda(TensorIterator& iter){
```
sign->sgn

mruberry commented on benchmarks/operator_benchmark/pt/unary_test.py:
```diff
@@ -94,6 +94,7 @@ def forward(self):
         ['sigmoid', torch.sigmoid],
         ['sigmoid_', torch.sigmoid_],
         ['sign', torch.sign],
+        ['sgn', torch.sign],
```
torch.sgn?

mruberry commented on aten/src/ATen/native/cuda/UnarySignKernels.cu:
```diff
@@ -45,8 +45,22 @@ void sign_kernel_cuda(TensorIterator& iter){
 
+template<typename T>
+__host__ __device__ static inline c10::complex<T> sgn_wrapper(c10::complex<T> v) {
+  return v.sgn();
```
Sorry, I'm not very familiar with c10::complex (yet), how does this work? Alternatively, where is this function implemented?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5376,6 +5376,28 @@ def merge_dicts(*dicts):
+""" + r"""
+Args:
+    {input}
+    {out}
```
Out is an optional kwarg

anjali411 replied:
yeah but all the other functions define `out` in a similar way in the doc. also `sgn(input, out=None) -> Tensor` this is indicative of the fact that out is optional kwarg I think

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5376,6 +5376,28 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> x=torch.randn(4, dtype=torch.cfloat)
```
The example doesn't really demonstrate the function because of the random values. Maybe pick some angles, like 0, 45, 90 degrees?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5376,6 +5376,28 @@ def merge_dicts(*dicts):
+           r"""
+sgn(input, out=None) -> Tensor
+
+Returns a new complex tensor with the same angle as that of the elements of :attr:`input` and absolute value 1.
```
Probably want to reference the angle and absolute value functions

mruberry replied:
Oh wait... we don't have complex abs yet, do we? Then you may need to define what you mean by "absolute value"

anjali411 replied:
yeah good idea. Btw both `abs` and `angle` are defined for complex tensors.

ezyang replied:
Reading over this I kind of wonder why we didn't call this `normalize` or `hat` since that's exactly what we're doing: we're treating the complex numbers as vectors and normalizing them into unit vectors. (OK, I do know why we're calling it sgn, it's because it's the Matlab sgn thing. But maybe in the docs it would be worth mentioning this.)

ezyang replied:
What does this function do when you have zero complex numbers in the input?

anjali411 replied:
it returns 0 (following https://en.wikipedia.org/wiki/Sign_function#Complex_signum)

mruberry commented on test/test_torch.py:
```diff
@@ -11128,6 +11128,19 @@ def test_sign(self, device):
+        cos_angle = angle.cos()
+        sin_angle = angle.sin()
+        expected = cos_angle + 1j * sin_angle
+        self.assertEqual(x.sgn(), expected)
```
Can this instead assert that x.sgn has the same angle as x.angle and that the absolute value is every one?

anjali411 replied:
yeah sure

mruberry commented:
Added some notes. This also still needs to update torch/_overrides.py, docs/source/tensors.rst, docs/source/torch.rst, docs/source/name_inference.rst, aten/src/ATen/core/aten_interned_strings.h and tools/derivatives.yaml.

Can this also be tested by TestTensorDeviceOps and TestTorchMathOps?

ezyang commented:
Relevant issues: https://github.com/pytorch/pytorch/issues/36323 https://github.com/pytorch/pytorch/issues/36655

ezyang commented on aten/src/ATen/native/cpu/zmath.h:
```diff
@@ -138,6 +138,16 @@ inline c10::complex<double> ceil_impl (c10::complex<double> z) {
 
+inline c10::complex<float> sgn_impl (c10::complex<float> z) {
+  auto angle = std::arg(z);
+  return c10::complex<float>(std::cos(angle), std::sin(angle));
```
I actually find this is a pretty unexpected way of computing the normalized vector. Ordinarily, I would have expected you compute the absolute value and then divide by that. Trigonometric functions are generally more expensive, and here you're computing THREE trig functions (atan2, cos, sin), whereas computing the norm and then dividing would just require a square, addition, square root and then division.

anjali411 replied:
I agree but this function is is a more numerically robust version of x/abs(x) (which is why we decided to add this function in the first place)

ngimel replied:
rsqrt, no division (if you don't care about precision)!

anjali411 committed:
Update on "torch.sgn for complex tensors"




[ghstack-poisoned]

ezyang reviewed (changes requested):
Don't compute normalized vector using trig functions

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

TODO:
1. add tests for backward (waiting on gradcheck PR for complex)

[ghstack-poisoned]

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5586,6 +5586,29 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.sgn,
+           r"""
+sgn(input, out=None) -> Tensor
```
input, *, out=None

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5586,6 +5586,29 @@ def merge_dicts(*dicts):
+
+""" + r"""
+Args:
+    {input}
```
```
Keyword args:
  {out}
```

mruberry reviewed (approved):
Minor doc nits but otherwise this looks good to me!

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

TODO:
1. add tests for backward (waiting on gradcheck PR for complex)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/cuda/UnarySignKernels.cu:
```diff
@@ -51,9 +51,26 @@ void signbit_kernel_cuda(TensorIterator& iter){
 }
 
+template<typename T>
+__host__ __device__ static inline c10::complex<T> sgn_wrapper(c10::complex<T> z) {
```
this could be an alternate way of implementing this function:
```
template<typename T>
__host__ __device__ static inline c10::complex<T> sgn_wrapper(c10::complex<T> z) {
  if (z == c10::complex<T>(0, 0)) {
    return c10::complex<T>(0, 0);
  } else {
    auto angle = std::arg(z);
    auto cos = std::cos(angle);
    auto mod_sin = std::sqrt(1 - cos*cos);
    auto sin = (z.real() > 0) ? mod_sin : mod_sin * -1;
    return c10::complex<T>(cos, sin);
  }
}
template<typename T>
__host__ __device__ static inline c10::complex<T> sgn_wrapper(c10::complex<T> z) {
  if (z == c10::complex<T>(0, 0)) {
    return c10::complex<T>(0, 0);
  } else {
    auto angle = std::arg(z);
    auto cos = std::cos(angle);
    auto mod_sin = std::sqrt(1 - cos*cos);
    auto sin = (z.real() > 0) ? mod_sin : mod_sin * -1;
    return c10::complex<T>(cos, sin);
  }
}
``` 

The performance for the two versions are:
x/ |x| :  

```
In [2]: x=torch.randn(100, dtype=torch.cfloat)

In [3]: %timeit x.sgn()
60.5 µs ± 1.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [4]: %timeit x.cuda().sgn()
250 µs ± 11.6 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

above mentioned implementation:

```
In [2]: x=torch.randn(100, dtype=torch.cfloat)

In [5]:  %timeit x.cuda().sgn()
242 µs ± 4.25 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [6]: %timeit x.cuda().sgn()
237 µs ± 5.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit x.cuda().sgn()
243 µs ± 5.44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

cc. @ngimel @ezyang

ngimel replied:
1) Please use `Timer` utility for cuda measurements, %timeit is not doing necessary synchronization https://github.com/pytorch/pytorch/tree/master/torch/utils/_benchmark. 
2) benchmark on the variety of sizes, benchmarking on the tensor of size 100 is not particularly informative. 
3) you are benchmarking time to transfer tensor to the GPU + time of the operation, that's probably not what you want.

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -274,6 +274,18 @@
+
+- func: sgn.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
+    CPU: sgn_out
```
Is this dispatch required or will the system automatically match these functions?

ngimel replied:
Dispatch is not required, `sgn_out` function is the same for CPU and CUDA backends, actual backend selection is handled by stubs.

anjali411 replied:
you are right, it's not needed. removed

mruberry commented on docs/source/name_inference.rst:
```diff
@@ -197,6 +197,8 @@ If you don't see an operation listed here, but it would help your use case, plea
    :meth:`Tensor.sigmoid_`,None
    ":meth:`Tensor.sign`, :func:`torch.sign`",:ref:`keeps_input_names-doc`
    :meth:`Tensor.sign_`,None
+   ":meth:`Tensor.sgn`, :func:`torch.sgn`",:ref:`keeps_input_names-doc`
```
You can revert this if you like since the native_function entries don't support named tensors and it's untested.

anjali411 replied:
synced offline - not removing it as other unary ops have their entry in this list as well.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6400,6 +6400,31 @@ def merge_dicts(*dicts):
+           r"""
+sgn(input, *, out=None) -> Tensor
+
+For complex tensors, this function returns a new tensor with the same angle as that of the
```
returns a new tensor whose elements have the same angle as those of :attr:\`input\` and an absolute value of one

mruberry reviewed (approved):
Still looks good. I made a few small suggestions.

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -773,6 +773,12 @@ Tensor trace_backward(const Tensor & grad, IntArrayRef sizes) {
 }
 
+Tensor sgn_backward(Tensor result, Tensor grad, Tensor self) {
+  // [grad / abs(z) - Re(grad/self) * result
```
I think there are mismatched parenthesis here no?
Also `z` is not defined.
Also the special case at 0 might need to be mentioned if you explain what this function is computing.

albanD replied:
For a real Tensor this should always return 0 right? That does not seem to be the case here when self < 0 right?

albanD reviewed (commented):
I think this still needs the gradients for the real case to be tested, even if you postpone the check for the complex version.

anjali411 commented:
> I think this still needs the gradients for the real case to be tested, even if you postpone the check for the complex version.
 
I was thinking of waiting to merge this PR until #43208 is merged and add tests for sgn.

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

This PR doesn't test the correctness of the gradients. It will be done as a part of auditing all the ops in future once we decide the autograd behavior (JAX vs TF) and add gradchek.

Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

also updates the backward definition of `torch.div`, `torch.abs`
Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

also updates the backward definition of `torch.div`, `torch.abs`
Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

also updates the backward definition of `torch.div`, `torch.abs`
Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

also updates the backward definition of `torch.div`, `torch.abs`
Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.sgn for complex tensors"


resolves https://github.com/pytorch/pytorch/issues/36323 by adding `torch.sgn` for complex tensors.
`torch.sgn` returns `x/abs(x)` for `x != 0` and returns `0 + 0j` for `x==0`

also updates the backward definition of `torch.div`, `torch.abs`
Differential Revision: [D23460526](https://our.internmc.facebook.com/intern/diff/D23460526)

[ghstack-poisoned]
